README.md
# Eyes-Testcafe

<img src="https://img.shields.io/badge/node-%3E%3D%2010-green">

Applitools Eyes SDK for [Testcafe](https://devexpress.github.io/testcafe/).

## Installation

Install Eyes-Testcafe as a local dev dependency in your tested project:

```bash
npm i -D @applitools/eyes-testcafe
```

## Applitools API key

In order to authenticate via the Applitools server, you need to supply the Eyes-Testcafe SDK with the API key you got from Applitools. Read more about how to obtain the API key [here](https://applitools.com/docs/topics/overview/obtain-api-key.html).

To to this, set the environment variable `APPLITOOLS_API_KEY` to the API key before running your tests.
For example, on Linux/Mac:

```bash
export APPLITOOLS_API_KEY=<your_key>
npx testcafe chrome:headless some-test-dir
```

And on Windows:

```bash
set APPLITOOLS_API_KEY=<your_key>
npx testcafe chrome:headless some-test-dir
```

It's also possible to specify the API key in the `applitools.config.js` file. The property name is `apiKey`. For example:

```js
module.exports = {
  apiKey: 'YOUR_API_KEY',
  ...
}
```

See the [Advanced configuration](#method-3-the-applitoolsconfigjs-file) section below for more information on using the config file.

## Usage

After defining the API key, you will be able to use commands from Eyes-Testcafe in your testcafe tests to take screenshots and use Applitools Eyes to manage them:

### Example

```js
import Eyes from '@applitools/eyes-testcafe';
const eyes = new Eyes();

fixture`Hello world`
  .page('https://applitools.com/helloworld')
  .afterEach(async () => eyes.close())
  .after(async () => eyes.waitForResults());
  
test('Cookies', async t => {
  await eyes.open({
      appName: 'Hello World!',
      testName: 'My first JavaScript test!',
      browser: [{ width: 800, height: 600, name: 'firefox' }],
      t
  });
  await eyes.checkWindow('Main Page');
  await t.click('button')
  await eyes.checkWindow('Click!');
});
```
<br/>

### Index
- [API](#API)
  - [Open](#open)
  - [CheckWindow](#checkWindow)
    - [tag](#tag)
    - [target](#target)
    - [fully](#fully)
    - [selector](#selector)
    - [region](#region)
    - [ignore](#ignore)
    - [floating](#floating)
    - [layout](#layout)
    - [strict](#strict)
    - [content](#content)
    - [accessibility](#accessibility)
    - [scriptHooks](#scriptHooks)
    - [sendDom](#sendDom)
  - [Close](#Close)
  - [waitForResults](#waitForResults)
  - [Best practices](#Best-practice-for-using-the-SDK)
- [Concurrency](#Concurrency)
  - [Running wih testcafe concurrency](#Running-wih-testcafe-concurrency)
  - [Applitools concurrency](#Applitools-concurrency)
- [Advanced configuration](#Advanced-configuration)
  - [Scoped configuration](#configuration-properties)
  - [Global configuration](#global-configuration-properties)
  - [Examples](#Method-1-Arguments-for-eyesOpen)
    - [Arguments for `eyesOpen`](#Method-1-Arguments-for-eyesOpen)
    - [Environment variables](#Method-2-Environment-variables)
    - [The `applitools.config.js` file](#Method-3-The-applitoolsconfigjs-file)
- [Configuring the browser](#Configuring-the-browser)
  - [Device emulation](#Device-emulation)
- [Troubleshooting](#Troubleshooting)

<br/><br/><br/>

# API

### **open**

Create an Applitools test.
This will start a session with the Applitools server.

```js
eyes.open({
  appName: '',
  testName: '',
  t
});
```

It's possible to pass a config object to `open` with all the possible configuration properties. Read the [Advanced configuration](#advanced-configuration) section for a detailed description.

### **checkWindow**

Generate a screenshot of the current page and add it to the Applitools Test.

```js
eyes.checkWindow(tag)

OR

eyes.checkWindow({ tag: 'your tag', target: 'your target mode' })
```

#### Arguments to `eyes.checkWindow`

- #### tag 
  (optional): A logical name for this check.
- #### target
   (optional): Possible values are:

  - **`window`**: This is the default value. Capture the entire window or only the viewport. If set then add [fully](#fully) as sibling to determine weather to capture full screen or viewport.
  - **`region`**: Take a screenshot of a region of the page, specified by coordinates or a selector. If set then add [region](#region) or [selector](#selector) as siblings for specifying the region/s.

- #### fully: 
  (optional) In case [target](#target) is `window`, determines wether to capture full page or viewport only. if `true` (default) then captures full page, if `false` then captures viewport.

    ```js
        // capture viewport only
        eyes.checkWindow({
          target: 'window',
          fully: false,
        });
    ```
- #### selector 
  (optional) In case [target](#target) is `region`, this should be the actual css, xpath or [a Testcafe Selector](https://devexpress.github.io/testcafe/documentation/test-api/selecting-page-elements/selectors/) to an element, and the screenshot would be the content of that element. For example:

  ```js
      // The shorthand string version defaults to css selectors
      eyes.checkWindow({
        target: 'region',
        selector: '.my-element'
      });

      // Using a css selector
      eyes.checkWindow({
        target: 'region',
        selector: {
          type: 'css',
          selector: '.my-element' // or '//button'
        }
      });

      // Using a Testcafe Selector
      import {Selector} from 'testcafe';
      eyes.checkWindow({
        target: 'region',
        selector: Selector('.my-region')
      });

      // Using an xpath selector
      eyes.checkWindow({
        target: 'region',
        selector: {
          type: 'xpath',
          selector: '//button[1]'
        }
      });
  ```

- #### region 
  (optional) In case [target](#target) is `region`, this should be an object describing the region's coordinates. For example:

    ```js
    eyes.checkWindow({
      target: 'region',
      region: {top: 100, left: 0, width: 1000, height: 200}
    });
    ```

- #### ignore 
  (optional): A single or an array of regions to ignore when checking for visual differences. For example:

    ```js
    eyes.checkWindow({
      ignore: [
        {top: 100, left: 0, width: 1000, height: 100},
        {selector: '.some-div-to-ignore'},
        {selector: Selector('.some-div')}
      ]
    });
    ```

- #### floating
  (optional): A single or an array of floating regions to ignore when checking for visual differences. More information about floating regions can be found in Applitools docs [here](https://help.applitools.com/hc/en-us/articles/360006915292-Testing-of-floating-UI-elements). For example:

    ```js
    eyes.checkWindow({
      floating: [
        {top: 100, left: 0, width: 1000, height: 100, maxUpOffset: 20, maxDownOffset: 20, maxLeftOffset: 20, maxRightOffset: 20},
        {selector: '.some-div-to-float', maxUpOffset: 20, maxDownOffset: 20, maxLeftOffset: 20, maxRightOffset: 20},
        {selector: Selector('.some-div'), maxUpOffset: 20, maxDownOffset: 20, maxLeftOffset: 20, maxRightOffset: 20}
      ]
    });
    ```

- #### layout
  (optional): A single or an array of regions to match as [layout level.](https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels) For example:
    ```js
    eyes.checkWindow({
      layout: [
        {top: 100, left: 0, width: 1000, height: 100},
        {selector: '.some-div-to-test-as-layout'},
        {selector: Selector('.some-div')}
      ]
    });
    ```

- #### strict
  (optional): A single or an array of regions to match as [strict level.](https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels) For example:
    ```js
    eyes.checkWindow({
      strict: [
        {top: 100, left: 0, width: 1000, height: 100},
        {selector: '.some-div-to-test-as-strict'},
        {selector: Selector('.some-div')}
      ]
    });
    ```

- #### content
  (optional): A single or an array of regions to match as [content level.](https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels) For example:
    ```js
    eyes.checkWindow({
      content: [
        {top: 100, left: 0, width: 1000, height: 100},
        {selector: '.some-div-to-test-as-content'},
        {selector: Selector('.some-div')}
      ]
    });
    ```

- #### accessibility
  (optional): A single or an array of regions for checking accessibility on. For example:

  ```js
  eyes.checkWindow({
    accessibility: [
      {accessibilityType: 'RegularText', selector: '.some-div'},
      {accessibilityType: 'RegularText', selector: Selector('.some-div-2')},
      {accessibilityType: 'LargeText', selector: '//*[@id="main"]/h1', type: 'xpath'},
      {accessibilityType: 'BoldText', top: 100, left: 0, width: 1000, height: 100},
    ]
  });
  ```

    Possible accessibilityType values are: `IgnoreContrast`,`RegularText`,`LargeText`,`BoldText` and `GraphicalObject`.

- #### accessibilityLevel
    (optional): The accessibility level to use for the screenshot. Possible values are `None`, `AA` and `AAA`.
    ```js
    cy.eyesCheckWindow({accessibilityLevel: 'AA'})
    ```
    
- #### scriptHooks 
  (optional): A set of scripts to be run by the browser during the rendering. It is intended to be used as a means to alter the page's state and structure at the time of rendering.
    An object with the following properties:
      - `beforeCaptureScreenshot`: a script that runs after the page is loaded but before taking the screenshot. For example:

        ```js
        eyes.checkWindow({
          scriptHooks: {
            beforeCaptureScreenshot: "document.body.style.backgroundColor = 'gold'"
          }
        })
        ```

- #### sendDom
  (optional): A flag to specify whether a capture of DOM and CSS should be taken when rendering the screenshot. The default value is true. This should only be modified to troubleshoot unexpected behavior, and not for normal production use.

    ```js
    eyes.checkWindow({sendDom: false})
    ```

### **close**

Close the applitools test and check that all screenshots are valid.

It is important to call this at the end of each test, symmetrically to `open`(or in `afterEach()`, see [Best practice for using the SDK](#best-practice-for-using-the-sdk)).

Close receives no arguments.

```js
eyes.close();
```

### **waitForResults**
Wait until all tests in the fixture are completed and return their results.
Note that if you don't wait for the tests to be completed then in case of a visual test failure, eyes cannot fail the fixture.
* it is recommended to wait for the results in the testcafe `after()` hook. 

```js
await eyes.waitForResults()
```

waitForResults receives `rejectOnErrors`,
* If `true` (default) and a visual test fails then reject with an `Error` (in case of a general error reject as well).
If the rejection is not handled then Testcafe fails the fixture.
* If `false` and a visual test fails then `waitForResults` resolves with an `Error`.
In case of a general `Error` reject with the `Error`. 

In case all the tests passed then waitForResults resolves with the test results.
<!-- resolves with an Array, each element in the array represents a test (testcafe test, i.e. open checkWindow and close) and is an Array of TestResults. Each TestResults is a visual test environnement. -->
___
<br/>

## Best practice for using the SDK

Every call to `eyes.open` and `eyes.close` defines a test in Applitools Eyes, and all the calls to `eyes.checkWindow` between them are called "steps". In order to get a test structure in Applitools that corresponds to the test structure in Testcafe, it's best to open/close tests in every `test` call. **You can use `afterEach` for calling `eyes.close()`**
 
After all tests are done you should call eyes.waitForResults, **you can use `after()` for calling `eyes.waitForResults`**, this is is done for two reasons:
1. to signal testcafe to wait until all the tests have been completed.
2. to obtain test results if needed.

```js
fixture`Hello world`
  .page('https://applitools.com/helloworld')
  .afterEach(async () => eyes.close());
  .after(async () => eyes.waitForResults())
```

Applitools will take screenshots and perform the visual comparisons in the background. Performance of the tests will not be affected during the test run, but there will be a small phase at the end of the test run that waits for visual tests to end.

* When running tests [concurrently](https://devexpress.github.io/testcafe/documentation/guides/basic-guides/run-tests.html#run-tests-concurrently) you should create an eyes instance for each test, see [Running wih testcafe concurrency.](#Running-wih-testcafe-concurrency)
___
<br/>

## Concurrency

### Running wih testcafe concurrency

Since Applitools tests are started by calling `eyes.open()` and ended by calling `eyes.close()` then when running Testcafe [concurrently](https://devexpress.github.io/testcafe/documentation/guides/basic-guides/run-tests.html#run-tests-concurrently) you should create an eyes instance per test, so you can call `eyes.open()` and `eyes.close()` on a per test basis, see example:

```js
fixture `hello world`.page`https://www.example.com/hello.html`

test('test 1', async t => {
  const eyes = new Eyes();
  await eyes.open({appName: 'TestCafe', testName: 'test 1', t});
  await eyes.checkWindow({tag: 'Page 1'});
  await eyes.close()
  await eyes.waitForResults()
});

test('test 2', async t => {
  const eyes = new Eyes();
  await eyes.open({appName: 'TestCafe', testName: 'test 2', t});
  await eyes.checkWindow({tag: 'Page 2'});
  await eyes.close()
  await eyes.waitForResults() 
});
```

* Note that now each test would result in a different Applitools [batch](#Configuration-properties:), if you want to keep all the tests in the same batch you can set [batchId:](#Configuration-properties:)
   
  ```js
  const batchId = `bid_${String(Math.random()).slice(2)}`
  fixture `hello world`.page`https://www.example.com/hello.html`
  test('test 1', async t => {
    const eye = new Eyes();
    await eyes.open({batchId, t ...});
    ...
  });

  test('test 2', async t => {
    const eyes = new Eyes();
    await eyes.open({batchId, t ...});
  ...
  });
  ```

### Applitools concurrency

The default level of concurrency for free accounts is `1`. This means that visual tests will not run in parallel during your tests, and will therefore be slow.
If your account does support a higher level of concurrency, it's possible to pass a different value by specifying it in the property `concurrency` in the applitools.config.js file (see [Advanced configuration](#advanced-configuration) section below).

If you are interested in speeding up your visual tests, contact sdr@applitools.com to get a trial account and faster tests with more concurrency.
___
<br/>

## Advanced configuration

There are 3 ways to specify test configuration:
1) Arguments to `eyes.open()`
2) Environment variables
3) The `applitools.config.js` file

The list above is also the order of precedence, which means that if you pass a property to `eyes.open` it will override the environment variable, and the environment variable will override the value defined in the `applitools.config.js` file.

#### Configuration properties:

| Property name             | Default value               | Description   |
| -------------             |:-------------               |:-----------   |
| `testName`                | undefined | The test name |
| `browser`                 | { width: 800, height: 600, name: 'chrome' } | The size and browser of the generated screenshots. This doesn't need to be the same as the browser that Testcafe is running. It could be a different size and also a different browser. For more info and possible values, see the [browser section below](#configuring-the-browser).<br/><br/>*Note: for best performance if possible eyes resizes the viewport to the given browser size.*|
| `batchId`                 | random                      | Provides ability to group tests into batches. Read more about batches [here](https://applitools.com/docs/topics/working-with-test-batches/how-to-group-tests-into-batches.html). |
| `batchName`               | The name of the first test in the batch                   | Provides a name to the batch (for display purpose only). |
| `baselineEnvName`         | undefined                   | The name of the environment of the baseline. |
| `envName`                 | undefined                   | A name for the environment in which the application under test is running. |
| `ignoreCaret`             | false                       | Whether to ignore or the blinking caret or not when comparing images. |
| `matchLevel`              | Strict                      | The method to use when comparing two screenshots, which expresses the extent to which the two images are expected to match. Possible values are `Strict`, `Exact`, `Layout` and `Content`. Read more about match levels [here](http://support.applitools.com/customer/portal/articles/2088359). |
| `baselineBranchName`      | undefined                   | The name of the baseline branch. |
| `parentBranchName`        | undefined                   | Sets the branch under which new branches are created. |
| `saveFailedTests`         | false                       | Set whether or not failed tests are saved by default. |
| `saveNewTests`            | false                       | Set whether or not new tests are saved by default. |
| `properties`              | undefined                   | Custom properties for the eyes test. The format is an array of objects with name/value properties. For example: `[{name: 'My prop', value:'My value'}]`. |
| `compareWithParentBranch` | false                       |  |
| `ignoreBaseline`          | false                       |  |
| `notifyOnCompletion`  | false | If `true` batch completion notifications are sent. |
<br/>
<!-- | `accessibilityLevel` | None | The accessibility level to use for the screenshots. Possible values are `None`, `AA` and `AAA`. | -->

### Global configuration properties:

The following configuration properties cannot be defined using the first method of passing them to `eyes.open`. They should be defined either in the `applitools.config.js` file or as environment variables.

| Property name             | Default value               | Description   |
| -------------             |:-------------               |:-----------   |
| `apiKey`                  | undefined                   | The API key used for working with the Applitools Eyes server. See more info in the [Applitools API key](#applitools-api-key) section above |
| `showLogs`                | false                       | Whether or not you want to see logs. Logs are written to the same output of the Testcafe process. <br/><br/>_Note that you can also use [DEBUG=eyes*](https://github.com/visionmedia/debug) for debugging._|
| `serverUrl`               | Default Eyes server URL     | The URL of Eyes server |
| `proxy`                   | undefined                   | Sets the proxy settings to be used in network requests to Eyes server. This can be either a string to the proxy URI, or an object containing the URI, username and password.<br/><br/>For example:<br/>`{url: 'https://myproxy.com:443', username: 'my_user', password: 'my_password', isHttpOnly: false}`<br/>or:<br/>`"https://username:password@myproxy.com:443"` |
| `isDisabled`              | false                       | If true, all api calls to Eyes-Testcafe are ignored. |
| `failTestcafeOnDiff`       | true                        | If true, then the Testcafe test fails if an eyes visual test fails. If false and an eyes test fails, then the Testcafe test does not fail. 
| `tapDirPath`              | undefined                   | Directory path of a results file. If set, then a [TAP](https://en.wikipedia.org/wiki/Test_Anything_Protocol#Specification) file is created in this directory, the tap file name is created with the name eyes-[\<ISO-DATE\>](https://en.wikipedia.org/wiki/ISO_8601)\.tap and contains the Eyes test results <br><br/> _Note that results are scoped per spec file, this means that the results file is created once for each fixture file)._|
| `concurrency`             | 1                           | The maximum number of tests that can run concurrently. The default value is the allowed amount for free accounts. For paid accounts, set this number to the quota set for your account. |
| `dontCloseBatches`| false | If true, batches are not closed for  [notifyOnCompletion](#advanced-configuration).|

### Method 1: Arguments for `eyes.open`

Pass a config object as the only argument. For example:

```js
eyes.open({
  appName: 'My app',
  batchName: 'My batch',
  ...
  // all other configuration variables apply
})
```

### Method 2: Environment variables

The name of the corresponding environment variable is in uppercase, with the `APPLITOOLS_` prefix, and separating underscores instead of camel case:

```js
APPLITOOLS_APP_NAME
APPLITOOLS_SHOW_LOGS
APPLITOOLS_BATCH_NAME
APPLITOOLS_CONCURRENCY
APPLITOOLS_SAVE_DEBUG_DATA
APPLITOOLS_BATCH_ID
APPLITOOLS_BATCH_NAME
APPLITOOLS_BASELINE_ENV_NAME
APPLITOOLS_ENV_NAME
APPLITOOLS_IGNORE_CARET
APPLITOOLS_IS_DISABLED
APPLITOOLS_MATCH_LEVEL
APPLITOOLS_MATCH_TIMEOUT
APPLITOOLS_BRANCH_NAME
APPLITOOLS_BASELINE_BRANCH_NAME
APPLITOOLS_PARENT_BRANCH_NAME
APPLITOOLS_SAVE_FAILED_TESTS
APPLITOOLS_SAVE_NEW_TESTS
APPLITOOLS_COMPARE_WITH_PARENT_BRANCH
APPLITOOLS_IGNORE_BASELINE
APPLITOOLS_SERVER_URL
APPLITOOLS_PROXY
APPLITOOLS_NOTIFY_ON_COMPLETION
```
<!-- APPLITOOLS_ACCESSIBILITY_LEVEL -->

### Method 3: The `applitools.config.js` file

It's possible to have a file called `applitools.config.js` at the same folder location as `.testcaferc.json`. _(The directory from which you run TestCafe. This is usually the project's root directory)_. <br>In this file specify the desired configuration, in a valid JSON format. For example:

```js
module.exports = {
  appName: 'My app',
  showLogs: true,
  batchName: 'My batch'
  ...
  // all other configuration variables apply
}
```
<br/>

## Configuring the browser

Eyes-Testcafe will take a screenshot of the page in the requested browser, the browser can be set in the `applitools.config.js` or by passing it to `eyes.open`.

Possible values are:

- `chrome`
- `firefox`
- `edgechromium`
- `edgelegacy`
- `ie10`
- `ie11`
- `safari`
- `chrome-one-version-back`
- `chrome-two-versions-back`
- `firefox-one-version-back`
- `firefox-two-versions-back`
- `safari-one-version-back`
- `safari-two-versions-back`
- `edgechromium-one-version-back`

### Previous browser versions

`*-one-version-back` and `*-two-versions-back` are relative to the version of the same browser. For example, if `chrome` refers to version 79, then `chrome-one-version-back` will be Chrome 78 and `chrome-two-versions-back` will be Chrome 77.

### Getting a screenshot of multiple browsers in parallel

It's also possible to send an array of browsers, for example:

```js
eyes.open({
  ...
  browser: [
    {width: 800, height: 600, name: 'firefox'},
    {width: 1024, height: 768, name: 'chrome'},
    {width: 1024, height: 768, name: 'ie11'}
  ]
}
```
**Note**: that if only a single browser is set, then Eyes-Testcafe changes the testcafe's browser viewport to that size.  

### Device emulation

To enable chrome's device emulation, it's possible to send a device name and screen orientation, for example:

```js
eyes.open({
  ...
  browser: {
    deviceName: 'iPhone X',
    screenOrientation: 'landscape',
    name: 'chrome' // optional, just to make it explicit this is browser emulation and not a real device. Only chrome is supported for device emulation.
  }
}
```

Possible values for screen orientation are `landscape` and `portrait`, and if no value is specified, the default is `portrait`.

The list of device names is taken from [chrome devtools predefined devices](https://raw.githubusercontent.com/chromium/chromium/0aee4434a4dba42a42abaea9bfbc0cd196a63bc1/third_party/blink/renderer/devtools/front_end/emulated_devices/module.json), and can be obtained by running the following command in a unix-based shell (installing [`jq`](https://stedolan.github.io/jq/) might be needed):

```sh
curl -s https://raw.githubusercontent.com/chromium/chromium/0aee4434a4dba42a42abaea9bfbc0cd196a63bc1/third_party/blink/renderer/devtools/front_end/emulated_devices/module.json | jq '.extensions[].device.title'
```

In addition, it's possible to use chrome's device emulation with custom viewport sizes, pixel density and mobile mode, by passing `deviceScaleFactor` and `mobile` in addition to `width` and `height`. For example:

```js
eyes.open({
  ...
  browser: {
    width: 800,
    height: 600,
    deviceScaleFactor: 3,
    mobile: true,
    name: 'chrome' // optional, just to make it explicit this is browser emulation and not a real device. Only chrome is supported for device emulation.
  }
}
```
___
## Troubleshooting

* If issues occur, DEBUG_SAVE=1 env variable can be set to save helpful information. The information will be saved under a folder named `.applitools` in the current working directory. This could be then used for getting support on your issue.
 * You can also use [DEBUG=eyes*](https://github.com/visionmedia/debug) for debugging.


History.md

v1.7.4 / 2020-05-18
===================

* 1.7.4 [[edc7f82](https://github.com/applitools/eyes-testcafe/commit/edc7f82b4d924cacca6f6d350dd49766ebe990dc)]
* updated processPage dist file [[1a4f818](https://github.com/applitools/eyes-testcafe/commit/1a4f8182607e456699e0db9b1001041d285634e0)]
* updated packages [[114bb9a](https://github.com/applitools/eyes-testcafe/commit/114bb9a0db64644a3bb4674f4dba96b6ad72fa83)]
* updated vgc: using url instead of 'source' [[69e090e](https://github.com/applitools/eyes-testcafe/commit/69e090efe4e200c4eb490465292f9947f94cabf1)]
* removed only [[c513d3e](https://github.com/applitools/eyes-testcafe/commit/c513d3ead8831206005aa43352cdccfefedc1e87)]
* removed unused referrer param [[d92456f](https://github.com/applitools/eyes-testcafe/commit/d92456f76390a6b956ba635c9c6b0d48de9202f3)]
* better readem js exmaple [[60d2487](https://github.com/applitools/eyes-testcafe/commit/60d2487beac5e9a70874c560e99580bda14372aa)]
* added concurrency notes to readem [[aaacb48](https://github.com/applitools/eyes-testcafe/commit/aaacb4842be566fb1751a1ffb33e8d35d01a648e)]
* 1.7.3 [[c81881d](https://github.com/applitools/eyes-testcafe/commit/c81881d543a865527dd9b544e24f52b10974bda4)]
* update vgc to get edgechromium in types [[1430176](https://github.com/applitools/eyes-testcafe/commit/1430176b0fcdd1bb9e96080d758786f952cbfc91)]
* 1.7.2 [[75abaf3](https://github.com/applitools/eyes-testcafe/commit/75abaf3b474f30eeec456d3290ebf896b1ac2901)]
* add edgechromium to readme [[99e54b8](https://github.com/applitools/eyes-testcafe/commit/99e54b8218b5808d58e7d6a9d1473c9ac737f85b)]
* 1.7.1 [[716a1a3](https://github.com/applitools/eyes-testcafe/commit/716a1a389e1349ca56cb39493deb36f6b683f723)]
* 1.7.0 [[5dac25a](https://github.com/applitools/eyes-testcafe/commit/5dac25ae4d0327daebfdc8633269a0e3332dd8fc)]
* update deps to get edgechromium [[7f66109](https://github.com/applitools/eyes-testcafe/commit/7f66109fc32eb10b06619ec800b472c13b707678)]
* changelog [[3e18f99](https://github.com/applitools/eyes-testcafe/commit/3e18f99eedea8f2b5db2dd3772ba6d47836f9a31)]
* 1.6.0 [[250f147](https://github.com/applitools/eyes-testcafe/commit/250f147cdf44451a5c36eb77108992fc3223089e)]
* updatde eyes-sdk-core to support setting the config file path with an env variable; APPLITOOLS_CONFIG_PATH. [[69c5491](https://github.com/applitools/eyes-testcafe/commit/69c54911cd040944a059784d9fa926af6d46abc1)]
* changelog [[27d475b](https://github.com/applitools/eyes-testcafe/commit/27d475b0f26c6c91779237343d9f586149ab3bbe)]

v1.5.14 / 2020-03-31
====================

* 1.5.14 [[beff433](https://github.com/applitools/eyes-testcafe/commit/beff4338a1356bab83eef6e0bb26d28a9cb1ecb9)]
* removed eyes-common dependency [[37e5078](https://github.com/applitools/eyes-testcafe/commit/37e5078f11fd1816a7fb16147a47878e4605be95)]
* fix changelog [[46bafe9](https://github.com/applitools/eyes-testcafe/commit/46bafe927577256cdff1ec99d119ae4d9e4b93a6)]
* changelog [[b1e27e4](https://github.com/applitools/eyes-testcafe/commit/b1e27e4468163d8a9237480733a0305c29f2051d)]

v1.5.13 / 2020-03-26
====================

* 1.5.13 [[2b6714e](https://github.com/applitools/eyes-testcafe/commit/2b6714e9891584e91a1ece807759c4720a07779b)]

v1.5.12 / 2020-03-26
====================

* 1.5.12 [[9bbeb85](https://github.com/applitools/eyes-testcafe/commit/9bbeb85a252c36e83efdafbea2a3224cf2d1b345)]
* fix process page css tree compilation error [[d2bc085](https://github.com/applitools/eyes-testcafe/commit/d2bc0855d220424297e6e05f1dde7ac4fbc03988)]

v1.5.11 / 2020-03-10
====================

* 1.5.11 [[788ec44](https://github.com/applitools/eyes-testcafe/commit/788ec44587af0e74af6fa952d14b83cb9c27bf0e)]

v1.5.10 / 2020-03-10
====================

* 1.5.10 [[5231c61](https://github.com/applitools/eyes-testcafe/commit/5231c612af85df4c9dddcc6ffd5247069ab3ab63)]
* bump ec, esc, vgc [[11cd927](https://github.com/applitools/eyes-testcafe/commit/11cd92767aa47fc5ac9b49ff673eb74901a413b7)]
* changelog [[ac53b7b](https://github.com/applitools/eyes-testcafe/commit/ac53b7b0a34d79857480819a75c238d4216251da)]

v1.5.9 / 2020-03-08
===================

* 1.5.9 [[865515f](https://github.com/applitools/eyes-testcafe/commit/865515fb9ec39ad15c35bca339045212d18b3755)]
* fix getting branching info from server [[fec8fd6](https://github.com/applitools/eyes-testcafe/commit/fec8fd66bd607783a23c99cef8f2edeacaeeb44a)]
* changelog [[eccf3c5](https://github.com/applitools/eyes-testcafe/commit/eccf3c53dee1416979567fdab8ef9a25f8c778a1)]

v1.5.8 / 2020-02-26
===================

* 1.5.8 [[e222a0c](https://github.com/applitools/eyes-testcafe/commit/e222a0c90e299fd8134a9520263d1db81ac0307f)]
* added branch base commit support [[488e739](https://github.com/applitools/eyes-testcafe/commit/488e739557586ba8c450c1d833901c470bc1d136)]
* updated render script [[3989a49](https://github.com/applitools/eyes-testcafe/commit/3989a49197275d5dca720dd065d9412e5d7d4a0f)]
* changelog [[fb08365](https://github.com/applitools/eyes-testcafe/commit/fb083650d7d5da30117e89fc71e9d9c4df352ebc)]

v1.5.7 / 2020-02-16
===================

* 1.5.7 [[e8192b9](https://github.com/applitools/eyes-testcafe/commit/e8192b9ea3814cd794bfd53d0f51eaa0a0a74d76)]

v1.5.6 / 2020-02-16
===================

* 1.5.6 [[e329f58](https://github.com/applitools/eyes-testcafe/commit/e329f58497e6ad28fb57e4dcf11c2c7ea43d47bc)]
* updated vgc to suport branching feature [[ba3f467](https://github.com/applitools/eyes-testcafe/commit/ba3f46797f98a0502a3b3fc809360977ae526de8)]

v1.5.5 / 2020-01-28
===================

* 1.5.5 [[2e71e8b](https://github.com/applitools/eyes-testcafe/commit/2e71e8bfd0cafbfcb8f45897b67674b5bd5d5b66)]
* update deps to get device emulation validation [[66f90ca](https://github.com/applitools/eyes-testcafe/commit/66f90cac974b4624a91649f274875fb1e47d16dc)]

v1.5.4 / 2020-01-21
===================

* 1.5.4 [[6c79c06](https://github.com/applitools/eyes-testcafe/commit/6c79c0662997a74f6cdd6f04d11c3d8ccbc8010a)]

ts/eyes.ts
/// <reference types="@applitools/visual-grid-client" />
/// <reference types="testcafe" />

declare namespace Eyes {
    namespace Testcafe {
      interface OpenOptions extends Omit<Eyes.Open.Options, 'accessibilitySettings'> {
        t: TestController
        accessibilityValidation: AccessibilitySettings
      }
  
      interface Selector {
        selector: SelectorAPI
      }
  
      interface FloatingSelector extends Selector {
        maxUpOffset: number
        maxDownOffset: number
        maxLeftOffset: number
        maxRightOffset: number
      }
  
      interface AccessibilitySelector extends Selector {
        accessibilityType: Eyes.Check.AccessibilityType
      }
  
      type RegionOrSelector = Selector|Eyes.Check.Region|Eyes.Check.Selector
      type FloatingRegionOrSelector = FloatingSelector|Eyes.Check.FloatingRegion|Eyes.Check.FloatingSelector
      type AccessibilityRegionOrSelector = AccessibilitySelector|Eyes.Check.AccessibilityRegion|Eyes.Check.AccessibilitySelector
  
      interface CheckOptions extends Omit<Eyes.Check.Options, 'ignore'|'floating'|'layout'|'content'|'strict'|'accessibility'|'selector'> {
        /**
         * A single or an array of regions to ignore when checking for visual differences.
         */
        ignore?:RegionOrSelector|RegionOrSelector[]
        
        /**
         * A single or an array of floating regions to ignore when checking for visual differences. 
         * More information about floating regions can be 
         * found in https://help.applitools.com/hc/en-us/articles/360006915292-Testing-of-floating-UI-elements .
         */
        floating?:FloatingRegionOrSelector|FloatingRegionOrSelector[]
  
        /**
         * A single or an array of regions to match as layout level.
         * See: https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels
         */
        layout?:RegionOrSelector|RegionOrSelector[]
  
        /**
         * A single or an array of regions to match as content level.
         * See: https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels
         */
        content?:RegionOrSelector|RegionOrSelector[]
  
        /**
         * A single or an array of regions to match as strict level.
         * See: https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels
         */
        strict?:RegionOrSelector|RegionOrSelector[]
  
        /**
         * A single or an array of regions to perform accessibility checks.
         */
        accessibility?:AccessibilityRegionOrSelector|AccessibilityRegionOrSelector[]

        /**
         * In case target is region, this should be the a css, xpath or a Testcafe Selector to an element, 
         * and the screenshot would be the content of that element.
         */
        selector?:SelectorAPI|string|Eyes.Check.Selector
      }
    }
  }